spec/GraphQLOverHTTP.md

```diff
@@ -187,94 +333,277 @@ This request could be sent via an HTTP GET as follows:
+If a literal `null` is desired, the parameter (e.g. {operationName}) should be
+omitted.
 
+GET requests MUST NOT be used for executing mutation operations. If the values
+of {query} and {operationName} indicate that a mutation operation is to be
+executed, the server MUST respond with error status code `405` (Method Not
+Allowed) and halt execution. This restriction is necessary to conform with the
+long-established semantics of safe methods within HTTP.
 
 ## POST
 
-A GraphQL POST request instructs the server to perform a query or mutation operation. A GraphQL POST request MUST have a body which contains values of the request parameters encoded according to the value of `Content-Type` header of the request.
+A GraphQL POST request instructs the server to perform a query or mutation
+operation. A GraphQL POST request MUST have a body which contains values of the
+request parameters encoded in one of the officially recognized GraphQL media
+types, or another media type supported by the server.
+
+A client SHOULD indicate the media type of a request body using the
+`Content-Type` header as specified in
+[RFC7231](https://datatracker.ietf.org/doc/html/rfc7231).
+
+Note: If the client does not include a `Content-Type` header then
+`application/json` will be assumed.
 
-A client MUST include a `Content-Type` with a POST request.
+### JSON Encoding
+
+When encoded in JSON, a GraphQL-over-HTTP request is a JSON object (map), with
+the properties specified by the GraphQL-over-HTTP request:
+
+- {query} - the string representation the Source Text of the Document as
+  specified in
+  [the Language section of the GraphQL specification](https://spec.graphql.org/draft/#sec-Language).
+- {operationName} - an optional string
+- {variables} - an optional object (map), the keys of which are the variable
+  names and the values of which are the variable values
+- {extensions} - an optional object (map)
+
+### Legacy watershed
+
+Before 1st January 2025 (`2025-01-01T00:00:00Z`), if a client does not know the
+media types the server supports then it SHOULD use
+`Content-Type: application/json` in requests. After this date, if a client does
+not know the media types the server supports then it SHOULD use
+`Content-Type: application/graphql+json` in requests.
 
 ### Example
 
-Given a POST request with a content type of `application/graphql+json` here is an example request body:
+If we wanted to execute the following GraphQL query:
+
+```graphql example
+query ($id: ID!) {
+  user(id: $id) {
+    name
+  }
+}
+```
+
+With the following query variables:
 
 ```json example
 {
-  "query": "query($id: ID!){user(id:$id){name}}",
+  "id": "QVBJcy5ndXJ1"
+}
+```
+
+This request could be sent via an HTTP POST to the relevant URL using the JSON
+encoding with the headers:
+
+```headers example
+Content-Type: application/graphql+json
+Accept: application/graphql+json
+```
+
+And the body:
+
+```json example
+{
+  "query": "query ($id: ID!) { user(id:$id) { name } }",
```

Comment:
```suggestion
  "query": "query($id:ID!){user(id:$id){name}}",
```

Or add in `\n` and whitespace to make it equivalent to the previous example

Reply:
I simplified the query first, but then decided it was valuable to show `\n` in the encoded data, so I have gone with the whitespace option :+1: Thanks @spawnia

Reply:
I agree that inclusion of whitespace makes it obvious that the string can be included as-is and does not need to be compressed. Should we recommend or mention compression in a non-normative section?